doc/designs/quic-design/server/quic-polling.md

```diff
@@ -0,0 +1,1637 @@
+retained mode interface, with option 2 being a fairly substantial hack and
+option 5 being the preferred approach for projects wanting an epoll/kqueue-style
+model on Windows. The suggested approach is therefore to implement option 5,
+though option 1 is also a viable fallback.
+
+In any case, it appears the poller API as designed and proposed above
+can be implemented adequately on Windows.
+
+Extra features on QUIC objects
+------------------------------
+
+### Low-watermark functionality
+
+Sometimes an application knows it does not need to do anything until at least N
+bytes are available to read or write. In conventional Berkeley sockets APIs this
+is known as “low-watermark” (LOWAT) functionality.
+
+Rather than making polling interfaces more convoluted by adding fields to
+polling-related structures, we propose to add a knob which can be configured on
+an individual QUIC stream:
+
+```c
+#define SSL_LOWAT_FLAG_ONESHOT     (1U << 0)
+
+int SSL_set_read_lowat(SSL *ssl, size_t loawt, uint64_t flags);
+int SSL_get_read_lowat(SSL *ssl, size_t *lowat);
+
+int SSL_set_write_lowat(SSL *ssl, size_t lowat, uint64_t flags);
+int SSL_get_write_lowat(SSL *ssl, size_t *lowat);
+```
+
+If `ONESHOT` is set, the low-watermark condition is automatically cleared
+after the next call to a read or write function respectively. The low-watermark
+condition can also be cleared by passing a low-watermark of 0.
+
+If low-watermark mode is configured, a poller will not report a stream as having
+data ready to read, or room to write data, if the amount of room available is
+less than the configured watermark.
+
+### Timeouts
+
+It is desirable to be able to cause blocking I/O operations to time out. For
+example, an application might want to perform a blocking read from a peer but
+not wait for a certain amount of time.
+
+We support this with a configurable timeout per each type of operation.
+
+```c
+/* All operations - defined as separate bit for forward ABI compatibility */
+#define SSL_OP_CLASS_ALL        (1U << 0)
+/* The timeout concerns reads. */
+#define SSL_OP_CLASS_R          (1U << 1)
+/* The timeout concerns writes. */
+#define SSL_OP_CLASS_W          (1U << 2)
+/* The timeout concetns accepts. */
+#define SSL_OP_CLASS_A          (1U << 3)
+/* The timeout concerns new stream creation (which may be blocked on FC). */
+#define SSL_OP_CLASS_N          (1U << 4)
+/* The timeout concerns connects. */
+#define SSL_OP_CLASS_C          (1U << 5)
+
+/*
+ * If set, t is a deadline (absolute time), otherwise it is a duration which
+ * starts whenever an operation is commenced.
+ */
+#define SSL_TIMEOUT_FLAG_DEADLINE    (1U << 0)
+
+/*
+ * Configure a timeout for one or more operation types. At least one operation
+ * type must be specified. If t is NULL, the timeout is unset for the given
+ * operation. This may be called multiple times to set different timeouts
+ * for different operations.
+ */
+int SSL_set_timeout(SSL *ssl, uint64_t operation,
```

Comment:
We already have an `SSL_set_timeout` function that does something different.

Reply:
Renamed &mdash; not that we are implementing this right now.